[v7r1] ProxyProvider docs (#4532)
DIRACGrid/DIRAC: docs/source/AdministratorGuide/Resources/proxyprovider.rst

.. _resourcesProxyProvider:

==============
ProxyProvider
==============

This resource type provides an interface to obtain proxy certificates using a user identifier. The following proxy providers are presented here: ``DIRACCA``, ``PUSP``.

DIRACCA proxy provider
----------------------

> ProxyProvider implementation for the proxy generation using local Certification Authority (CA) credentials. DIRACCA type of the proxy provider is a simple CA, its main purpose is to generate a proxy on the fly for DIRAC users who do not have a certificate registered in DIRAC registry. To use it need to have CA certificate and key locally. Here are two ways to set up this type of proxy provider in the DIRAC configuration:

Reviewer: ```suggestion
ProxyProvider implementation for the proxy generation using local Certification Authority (CA) credentials. DIRACCA type of the proxy provider is a simple CA, its main purpose is to generate a proxy on the fly for DIRAC users who do not have a certificate registered in DIRAC registry. To use it needs to have CA certificate and key locally. Here are two ways to set up this type of proxy provider in the DIRAC configuration:
```

> Recomended way to set all required parameters in the DIRAC configuration::

Reviewer: ```suggestion
Recommended way to set all required parameters in the DIRAC configuration::
```

* ``Supplied``: the distinguished name fields list that must be present. If this parameter is not specified, the default value will be a ``CN``.
* ``Optional``: the distinguished name fields list that are allowed, but not required. If this parameter is not specified, the default value will be a ``C, O, OU, emailAddress``.
* ``DNOrder``: order of the distinguished name fields in a created user certificate. If this parameter is not specified, the default value will be a ``C, O, OU, CN, emailAddress``.
* ``<One of the supported distinguished names>``: to set default value for distinguished name field, for example.: ``OU = MY, DIRAC``.

Also, as an additional feature, this class can read properties from a simple openssl CA configuration file. To do this, just set the path to an existing openssl configuration file as a CAConfigFile parameter. Required parameters in the DIRAC configuration for its implementation::

    Resources
    {
      ProxyProviders
      {
        CFG_DIRAC_CA
        {
          ProviderType = DIRACCA
          CAConfigFile = /opt/dirac/pro/etc/openssl_config_ca.cnf
          DNOrder = O, OU, CN, emailAddress

>       }

Author: I'll add a mention about DIRAC installations that don't need to use `ProxyProviders` resources.
This is related to OAuth extension as follows. Through OAuth2, we can also access some resources, such as [RCAuth CA] (https://rcauth.eu/), which is a service that can provide proxies through OAuth authorization.

And I forgot to describe one more important thing: with the implementation of `ProxyProvider` resources, DN may have a proxyProvider property, which should be specified in the `Registry/Users/<username>/` configuration section, if is necessary to use this` ProxyProvider` resource for current DN.
I'll add this description also.

.. _resourcesProxyProvider:

==============
ProxyProvider
==============

This resource type provide interface to obtain proxy certificate using user indentifier. The following proxy providers are represented here: ``DIRACCA``, ``PUSP``.

DIRACCA proxy provider
----------------------

> ProxyProvider implementation for the proxy generation using local CA credentials. This class is a simple CA, it's main purpose is to generate a simple proxy on the fly for DIRAC users who do not have any certificate register. To use it need to have CA certificate and key localy. Required parameters in the DIRAC configuration for its implementation::

Reviewer: ```suggestion
ProxyProvider implementation for the proxy generation using local Certification Authority (CA) credentials. This class is a simple CA, its main purpose is to generate a proxy on the fly for DIRAC users who do not have a certificate registered in DIRAC registry. To use it need to have CA certificate and key locally. Required parameters in the DIRAC configuration for its implementation::
```

But TBH I am not fully sure I understand what it does...?

Author: Thanks for your corrections, I'll think of a way to explain more clearly here.

    Resources
    {
      ProxyProviders
      {
        MY_DIRAC_CA
        {
          ProviderType = DIRACCA
          CertFile = /opt/dirac/pro/etc/grid-security/MY_DIRAC_CA.pem
          KeyFile = /opt/dirac/pro/etc/grid-security/MY_DIRAC_CA.key.pem
          Match = O, OU
          Supplied = CN
          Optional = emailAddress
          DNOrder = O, OU, CN, emailAddress
        }
      }
    }

Configuration options are:

* ``ProviderType``: main option, to show which proxy provider type you want to register. In the case of the DIRACCA proxy provider you must type ``DIRACCA``.
* ``CertFile``: the path to the CA certificate.
* ``KeyFile``: the path to the CA key.
* ``CAConfigFile``: the path to the openssl configuration file. This is optional and not recomended to use. But if you choose to use this option, it is recommended to use a relatively simple configuration.
* ``Match``: the distinguished name fields that must contain the exact same contents as that field in the CA's DN.
* ``Supplied``: the distinguished name fields list that must be present.
* ``Optional``: the distinguished name fields list that are allowed, but not required.

> * ``DNOrder``: order of the distinguished name fields in a created user certificate.

Reviewer: Are all these parameters needed, or they are anyway the defaults?

Author: good point,
Here are two ways to set up this type of proxy provider:
1. all the necessary parameters set in CS:
    all of these options are required, but each has a default value if it will not set in CS.
2. set in CS only the existing path of the openssl configuration file `CAConfigFile` and `DNOrder` is optional here:
    in this case all parameters will be read from this file.

So I'll add this to the description.

* ``<One of the supported distinguished names>``: to set default value for distinguished name field, for example.: ``OU = MY, DIRAC``.

Also, as an additional feature, this class can read properties from a simple openssl CA configuration file. To do this, just set the path to an existing openssl configuration file as a CAConfigFile parameter. In this case, the distinguished names order in the created proxy will be the same as in the configuration file policy block.

> The Proxy provider supports the following distinguished names (more details about distinguished names here: https://www.cryptosys.net/pki/manpki/pki_distnames.html):

Reviewer: This is normally not the way you specify links, check for some examples in existing RST files.

> * SN(surname)

Reviewer 2: Need an empty line before the bullet list
```suggestion

* SN(surname)
```

Author: done
